rust_src/src/keymap.rs

```diff
@@ -62,6 +65,193 @@ pub extern "C" fn set_where_is_cache(val: LispObject) {
+                            localObject = menu_item_eval_property(filter);
+                        }
+                        break;
+                    }
+                }
+            } else {
+                // Return since object must be an invalid keymap
+                break localObject;
+            }
+        }
+        /* If the keymap contents looks like (STRING . DEFN), use DEFN.
+        Keymap alist elements like (CHAR MENUSTRING . DEFN)
+        will be used by HierarKey menus.  */
+        else if consObject.cdr().is_string() {
+            localObject = consObject.into();
+        } else {
+            break localObject;
+        }
+    }
+}
+
+// TODO: Fix this mess with all the unwrap but how?
+// TODO: Avoid using FCONS?
+#[no_mangle]
+pub extern "C" fn _copy_keymap_item(elt: LispObject) -> LispObject {
+    if elt.is_not_cons() {
+        elt;
+    }
+
+    // Turn elt into a cons cell in order to use its car
+    let mut elt = elt.as_cons().unwrap();
+    let mut res = elt;
+    let mut tem = elt;
+
+    // Check if this is a new format menu item
+    if tem.car().eq(Qmenu_item) {
+        // Copy the cell with menu-item marker
+        unsafe {
+            res = Fcons(tem.car(), tem.cdr()).as_cons().unwrap();
+            elt = Fcons(tem.car(), tem.cdr()).as_cons().unwrap();
+
+            // Check if the next cell is a cons
+            if elt.cdr().is_cons() {
+                // Copy the cell with menu-item name
+                elt.set_cdr(Fcons(tem.car(), tem.cdr()));
+                elt = elt.cdr().as_cons().unwrap();
+                tem = elt.cdr().as_cons().unwrap();
+            }
+        }
+    } else {
+        if tem.car().is_string() {
+            unsafe {
+                res = Fcons(tem.car(), tem.cdr()).as_cons().unwrap();
+                elt = Fcons(tem.car(), tem.cdr()).as_cons().unwrap();
+                // Copy the cell since copy-alist iddn't go this deep
+                if tem.car().is_string() {
+                    elt.set_cdr(Fcons(tem.car(), tem.cdr()));
+                    elt = elt.cdr().as_cons().unwrap();
+                    tem = elt.cdr().as_cons().unwrap();
+                }
+                if tem.car().eq(Qkeymap) {
+                    elt.set_cdr(copy_keymap(tem.into()));
+                } else if tem.car().eq(Qkeymap) {
+                    res = copy_keymap(elt.into()).as_cons().unwrap();
+                }
+            }
+        }
+    }
+    res.into()
+}
+
+/* Return the offset of POSITION, a click position, in the style of
+the respective argument of Fkey_binding.  */
+pub extern "C" fn _click_position(position: LispObject) -> ptrdiff_t {
+    let curr_pos: EmacsInt;
+
+    if position.is_integer() {
+        curr_pos = position.into();
+    } else if position.is_marker() {
+        curr_pos = marker_position(position).try_into().unwrap();
+    } else {
+        curr_pos = ThreadState::current_buffer_unchecked()
+            .get_pt()
+            .try_into()
```

Comment:
try_into without actually handling failures is something we would like to avoid. This can lead to Remacs crashing unexpectedly.